operator/helper/trace.go

```diff
@@ -0,0 +1,78 @@
+§// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package helper
+
+import (
+	"encoding/hex"
+	"fmt"
+	"github.com/open-telemetry/opentelemetry-log-collection/entry"
+	"github.com/open-telemetry/opentelemetry-log-collection/operator"
+)
+
+// NewTimeParser creates a new time parser with default values
+func NewTraceParser() TimeParser {
+	return TimeParser{
+	}
+}
+
+// TimeParser is a helper that parses time onto an entry.
+type TraceParser struct {
+	ParseTraceFrom  *entry.Field `mapstructure:"parse_trace_from,omitempty"  json:"parse_trace_from,omitempty"  yaml:"parse_trace_from,omitempty"`
+	PreserveTraceTo *entry.Field `mapstructure:"preserve_trace_to,omitempty" json:"preserve_trace_to,omitempty" yaml:"preserve_trace_to,omitempty"`
+	ParseSpanFrom  *entry.Field `mapstructure:"parse_span_from,omitempty"  json:"parse_span_from,omitempty"  yaml:"parse_span_from,omitempty"`
+	PreserveSpanTo *entry.Field `mapstructure:"preserve_span_to,omitempty" json:"preserve_span_to,omitempty" yaml:"preserve_span_to,omitempty"`
+	ParseFlagsFrom  *entry.Field `mapstructure:"parse_flags_from,omitempty"  json:"parse_flags_from,omitempty"  yaml:"parse_flags_from,omitempty"`
+	PreserveFlagsTo *entry.Field `mapstructure:"preserve_flags_to,omitempty" json:"preserve_flags_to,omitempty" yaml:"preserve_flags_to,omitempty"`
+}
+
+// Validate validates a TraceParser, and reconfigures it if necessary
+func (t *TraceParser) Validate(context operator.BuildContext) error {
+	if t.ParseTraceFrom == nil {
+		return fmt.Errorf("missing required parameter 'parse_trace_from'")
+	}
+	if t.ParseSpanFrom == nil {
+		return fmt.Errorf("missing required parameter 'parse_span_from'")
+	}
+
+	return nil
+}
+
+func parse(entry *entry.Entry,field *entry.Field, to *entry.Field) []byte {
+	value, ok := entry.Delete(field)
+	if !ok {
+		return nil
+	}
+
+	data, err := hex.DecodeString(fmt.Sprintf("%v",value))
```

Comment:
I don't have a lot of experience working with traces, so let me ask a very basic question:

Is it a safe assumption that `TraceID`, `SpanID`, and `TraceFlags` should always be hex encoded? In other words, is this implementation specific to any particular tracing technology or standard?

Reply:
Hex-encoding requirement and the specific names of the fields is part of Otel spec: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/overview.md#trace-context-in-legacy-formats

Reply:
So this implementation will allow us to parse tracing data from logs that adhere to the OTel spec. 

Certainly this is an improvement upon what is supported today, but isn't the purpose of the library to ingest logs from all sources? 

I'm just trying to understand whether we are solving this in a generalized way, or if we are making an assumption about log format. Sounds like it's the latter, in which case I think we should consider this implementation preliminary, and plan to generalize it later. 

Maybe something like:

```
filelog:
    include: [ /.../logs/*.jsonl ]
    operators:
      - type: "json_parser"
        timestamp:
          parse_from: "timestamp_ms"
          layout_type: "epoch"
          layout: "ms"
        severity:
          parse_from: "level"
        trace_id:
          parse_from: "trace_id"
          encoding: "hex" // default is "hex", for now hardcoded as "hex"
        span_id:
          parse_from: "span_id"
          encoding: "hex" // default is "hex", for now hardcoded as "hex"
        trace_flags:
          parse_from: "trace_flags"
          encoding: "hex" // default is "hex", for now hardcoded as "hex"
```

Reply:
I would even go so far as to say that the config in this PR needs to be changed in order to accomodate the above. We don't necessarily need to support the `encoding` parameter yet, but we should anticipate that it will be added later:

```
filelog:
    include: [ /.../logs/*.jsonl ]
    operators:
      - type: "json_parser"
        timestamp:
          parse_from: "timestamp_ms"
          layout_type: "epoch"
          layout: "ms"
        severity:
          parse_from: "level"
        trace_id:
          parse_from: "trace_id"
        span_id:
          parse_from: "span_id"
        trace_flags:
          parse_from: "trace_flags"
```

Reply:
@djaglowski I agree. Let's just make the defaults compliant with Otel spec.

Reply:
Mmm, in the suggested config the trace/span/flags are split, and if I understand it correctly that would mean 3 different operators? That also means that if you want to default you need to specify:

3 operators for defaults:

```
filelog:
    include: [ /.../logs/*.jsonl ]
    operators:
      - type: "json_parser"
        trace_id: {}
        span_id: {}
        trace_flags: {}
```

1 operator for defaults:

```
filelog:
    include: [ /.../logs/*.jsonl ]
    operators:
      - type: "json_parser"
        trace: {}
```

Anyone strong opinions?

Reply:
I see your point about what I suggested. I think it can be done with 1 operator though, by pushing it down one level, as shown in your last suggestion:
```yaml
filelog:
    include: [ /.../logs/*.jsonl ]
    operators:
      - type: json_parser
        trace: {}
```

Which, accounting for default values would give an effective configuration:
```yaml
filelog:
    include: [ /.../logs/*.jsonl ]
    operators:
      - type: json_parser
        trace:
          trace_id:
            parse_from: trace_id
          span_id:
            parse_from: span_id
          trace_flags:
            parse_from: trace_flags
```

One could partially customize this, as in:
```yaml
filelog:
    include: [ /.../logs/*.jsonl ]
    operators:
      - type: json_parser
        trace:
          span_id:
            parse_from: my_custom_span_id
```

And this would also allow us to extend as necessary in the future:
```yaml
filelog:
    include: [ /.../logs/*.jsonl ]
    operators:
      - type: json_parser
        trace:
          trace_id:
            parse_from: trace_id
            encoding: hex
          span_id:
            parse_from: my_custom_span_id
            encoding: md5
          trace_flags:
            parse_from: my_trace_flags_array
            encoding: array
            some_other_config_thing: my_value
```

The above may need some tweaking to make sense for traces, but my primary concern is ensuring that we can extend the design later to support trace parsing for non-OTel compliant logs.

Reply:
Alright, that makes sense. I'll do it like that (skipping the customization for now, defaulting to the otel spec)